File: runners/go/runner.go
package main

import (
	"bufio"
	"encoding/json"
	"flag"
	"fmt"
	"net"
	"os"
	"os/exec"
	"path/filepath"
)

type FileSpec struct {
	Path    string `json:"path"`
	Content string `json:"content"`
	Mode    int    `json:"mode"`
}
type Request struct {
	Method    string            `json:"method"`
	TaskID    string            `json:"task_id"`
	Files     []FileSpec        `json:"files"`
	TargetDir string            `json:"target_dir"`
	Cmd       string            `json:"cmd"`
	Cwd       string            `json:"cwd"`
	Port      int               `json:"port"`
	Env       map[string]string `json:"env"`
	LogPath   string            `json:"log_path"`
	PID       int               `json:"pid"`
	Host      string            `json:"host"`
}

func outOk(data any) {
	b, _ := json.Marshal(map[string]any{"ok": true, "version": "v1", "data": data})
	fmt.Println(string(b))
}
func outErr(code, msg string) {
	b, _ := json.Marshal(map[string]any{"ok": false, "version": "v1", "error": map[string]string{"code": code, "message": msg}})
	fmt.Println(string(b))
}

func main() {
	jsonArg := flag.String("json", "", "json request")
	flag.Parse()
	var req Request
	if *jsonArg != "" {
		_ = json.Unmarshal([]byte(*jsonArg), &req)
	} else {
		_ = json.NewDecoder(os.Stdin).Decode(&req)
	}

	switch req.Method {
	case "apply":
		target := req.TargetDir
		if target == "" {
			target = "."
		}
		writes := 0
		for _, f := range req.Files {
			p := filepath.Join(target, f.Path)
			_ = os.MkdirAll(filepath.Dir(p), 0o755)
			_ = os.WriteFile(p, []byte(f.Content), 0o644)
			writes++
		}
		outOk(map[string]any{"writes": writes, "target": target})
	case "start":
		env := os.Environ()
		env = append(env, fmt.Sprintf("PORT=%d", req.Port))
		for k, v := range req.Env {
			env = append(env, fmt.Sprintf("%s=%s", k, v))
		}
		if req.Cwd == "" {
			req.Cwd = "."
		}
		cmd := exec.Command("bash", "-lc", req.Cmd)
		cmd.Dir = req.Cwd
		cmd.Env = env
		stdout, _ := cmd.StdoutPipe()
		stderr, _ := cmd.StderrPipe()
		if err := cmd.Start(); err != nil {
			outErr("E_RUNTIME", err.Error())
			return
		}
		if req.LogPath != "" {
			go func() {
				f, _ := os.OpenFile(req.LogPath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
				defer f.Close()
				r := bufio.NewReader(stdout)
				for {
					b, _, e := r.ReadLine()
					if e != nil {
						break
					}
					f.Write(append(b, '\n'))
				}
			}()
			go func() {
				f, _ := os.OpenFile(req.LogPath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
				defer f.Close()
				r := bufio.NewReader(stderr)
				for {
					b, _, e := r.ReadLine()
					if e != nil {
						break
					}
					f.Write(append(b, '\n'))
				}
			}()
		}
		outOk(map[string]any{"pid": cmd.Process.Pid})
	case "stop":
		p, _ := os.FindProcess(req.PID)
		_ = p.Kill()
		outOk(map[string]any{"stopped": true})
	case "health":
		h := req.Host
		if h == "" {
			h = "127.0.0.1"
		}
		c, e := net.Dial("tcp", fmt.Sprintf("%s:%d", h, req.Port))
		if e == nil {
			_ = c.Close()
			outOk(map[string]any{"ready": true})
		} else {
			outOk(map[string]any{"ready": false})
		}
	default:
		outErr("E_METHOD", "unknown method")
	}
}

File: runners/node/runner.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');
const net = require('net');

function outOk(data) {
  process.stdout.write(JSON.stringify({ ok: true, version: 'v1', data }) + '\n');
}

function outErr(code, message) {
  process.stdout.write(
    JSON.stringify({ ok: false, version: 'v1', error: { code, message } }) + '\n',
  );
}

function parseRequest() {
  const flagIndex = process.argv.indexOf('--json');
  if (flagIndex >= 0 && flagIndex + 1 < process.argv.length) {
    try {
      return JSON.parse(process.argv[flagIndex + 1] || '{}');
    } catch (err) {
      outErr('E_JSON', String(err && err.message ? err.message : err));
      process.exit(0);
    }
  }
  try {
    return JSON.parse(fs.readFileSync(0, 'utf8') || '{}');
  } catch (err) {
    outErr('E_JSON', String(err && err.message ? err.message : err));
    process.exit(0);
  }
}

function handleApply(req) {
  try {
    const target = req.target_dir || process.cwd();
    let writes = 0;
    for (const file of req.files || []) {
      const filePath = path.join(target, file.path);
      fs.mkdirSync(path.dirname(filePath), { recursive: true });
      fs.writeFileSync(filePath, file.content ?? '', 'utf8');
      if (typeof file.mode === 'number') {
        fs.chmodSync(filePath, file.mode);
      }
      writes += 1;
    }
    outOk({ writes, target });
  } catch (err) {
    outErr('E_FS', String(err && err.message ? err.message : err));
  }
}

function handleStart(req) {
  try {
    const cwd = req.cwd || process.cwd();
    const env = { ...process.env, ...(req.env || {}), PORT: String(req.port || 0) };
    const brewPaths = '/opt/homebrew/bin:/usr/local/bin';
    env.PATH = `${brewPaths}:${env.PATH || ''}`;

    const logPath = req.log_path || path.join(cwd, 'run.log');
    fs.mkdirSync(path.dirname(logPath), { recursive: true });
    fs.appendFileSync(logPath, `[runner] starting cmd: ${req.cmd} in ${cwd}\n`);

    const child = spawn('bash', ['-lc', req.cmd], {
      cwd,
      env,
      detached: true,
      stdio: ['ignore', 'pipe', 'pipe'],
    });

    child.stdout.on('data', (chunk) => fs.appendFileSync(logPath, chunk));
    child.stderr.on('data', (chunk) => fs.appendFileSync(logPath, chunk));
    child.on('error', (err) =>
      fs.appendFileSync(logPath, `[runner] child error: ${String(err)}\n`),
    );
    child.on('exit', (code) =>
      fs.appendFileSync(logPath, `[runner] child exit code: ${code}\n`),
    );

    child.unref();
    outOk({ pid: child.pid });
  } catch (err) {
    outErr('E_RUNTIME', String(err && err.message ? err.message : err));
  }
}

function handleStop(req) {
  try {
    process.kill(Number(req.pid), 'SIGTERM');
    outOk({ stopped: true });
  } catch (err) {
    outErr('E_RUNTIME', String(err && err.message ? err.message : err));
  }
}

function handleHealth(req) {
  const host = req.host || '127.0.0.1';
  const port = Number(req.port || 0);
  const socket = net.createConnection({ host, port }, () => {
    socket.end();
    outOk({ ready: true });
  });
  socket.on('error', () => outOk({ ready: false }));
  socket.setTimeout(1000, () => {
    socket.destroy();
    outOk({ ready: false });
  });
}

const request = parseRequest();

switch (request.method) {
  case 'apply':
    handleApply(request);
    break;
  case 'start':
    handleStart(request);
    break;
  case 'stop':
    handleStop(request);
    break;
  case 'health':
    handleHealth(request);
    break;
  default:
    outErr('E_METHOD', 'unknown method');
}
